### DISCUSSION TITLE
Adding support for lang_data.zip file operations on Audi vehicles

### DISCUSSION THREAD

**Discussion by filippz** [Mr-MIBonk/M.I.B._More-Incredible-Bash#441] [Ideas]
I have written a small app (.NET) that is able to use existing `lang_data.zip` **and add a language to it and/or change existing translations**.
To transfer `lang_data.zip` file itself I was using direct telnet connection to transfer the file from the unit to SD card and once changed I used the same way to copy it back to the unit.
**IMHO those operations could be simple added to the MIB tool in the following manner:**

- _backup of the unit could include `lang_data.zip` (`/net/mmx/mnt/app/eso/hmi/lsd/jars/lang_data.zip`)_
- _the MIB could offer the possibility of transferring `lang_data.zip` back to unit from the SD card (and reboot to apply)_

### Things to consider:

- instead of simply transferring the file back to unit script could also check the content of the `checksum_lang_data.md5` inside the zip file to confirm these are the same (as `lang_data.zip` varies over firmware versions `checksum_lang_data.md5` is used to verify that the `lang_data.zip` matches with firmware)
- if the user adds another language with the mentioned app there is also a need to add that language to the `DATASET_2D00` in `setlang` script (description of the mentioned dataset: https://mibwiki.one/doc/language-dataset-ggPAc1Or7X)
  
**Any thoughts about this? Anybody willing to test the app and take a crack at integration with MIB tool?**

### Additional notes:

- I have only used my own MHI2 unit in my Audi A4 B9 so I can't really tell if other units might benefit from the tool apart MHI2/MHI2Q in Audi vehicles. In any case more testing is needed.
- Details of the format have been described here: https://mibwiki.one/doc/audi-mhi2mhi2q-language-files-hd6YoHG0fx/insights
- Some translations contain placeholders (ie `%1`) which should be kept but the app doesn't force it.
- App loads the `lang_data.zip` and tries to generate it back from it's internal representation of it and checks if there are any differences - if there are differences then the save function is disabled in order to prevent the generation of the invalid file.
- App has a possibility to use "reference" `lang_data.zip` to speed up translations in the following manner – for example MHI2_ER_AUG22 doesn't have `ro_RO` language but `MHI2Q_ER_AUG22` does have it and both have `en_GB`:
1. Open the MHI2 `lang_data.zip`
2. Select language `en_GB` from it to use as a base for translation and add a new `ro_RO` language as a copy of the `en_GB`. Now we have `ro_RO` in MHI2 `lang_data.zip` that is exactly the same as the `en_GB`
3. Open the `lang_data.zip` from MHI2Q that already contains `en_GB` and `ro_RO` and select `en_GB` as a reference
5. App tries to find the matching translations from MHI2 by searching for it in MHI2Q `en_GB` (reference language) and if it finds it then it places the translation next to the untranslated MHI2 string from `ro_RO` (selected language) translation in MHI2Q
6. User has a possibility to either copy the offered translation to MHI2 and/or translate the message manually
![image_2024-03-26_12-45-52](https://github.com/Mr-MIBonk/M.I.B._More-Incredible-Bash/assets/4645804/4967f8ce-e5f2-4772-a3a4-e2084e7591a0)

**Comment by duxtodor-bit** [comment]
Greetings Filip, 
I like this idea, I have an MHI2Q device and 
I would be happy to do a language translation if I could get your translation tool

**Reply by filippz** [reply]
Attached you'll find the app coupled with two `lang_data` files. One is MHI2 (which for example doesn't have `sl_SI`, ...) and the other is from MHI2Q (which for example has `sl_SI`).
If editing existing language then you just open the `lang_data` which you need, but if you want to add a new language to the you can add it as as copy of the selected language (it must be in the format language_COUNTRY) 
Additionally, after adding a language you can use "reference" `lang_data` to speed up the process by selecting another `lang_data` that already has the language you need and both `lang_data` files have common language (the language you used to add new language)
The "reference mode" is visible in the screenshot where I've added `hr_HR` language as a copy of `en_GB` from MHI2 and using reference language (`en_GB` because I used `en_GB` as base for adding `hr_HR`) to fetch translation to `sl_SI` from MHI2 (the closest of available languages from MHI2Q to `hr_HR`)
Adding completely new language would also require modifying (at least) the appropriate dataset: https://github.com/Mr-MIBonk/M.I.B._More-Incredible-Bash/blob/main/apps/setlang#L35

<img width="1473" height="393" alt="image" src="https://github.com/user-attachments/assets/c1c01a81-0f00-4f31-9978-407feae5579f" />

In the end app remained as a proof of concept and I haven't used it for actual translations.

[AudiMHI2Translator.zip](https://github.com/user-attachments/files/31211814/AudiMHI2Translator.zip)

**Comment by duxtodor-bit** [comment]
Thanks Filip, 
I think it won't be a problem to do this with your app, my software version is MHI2Q AUG22_P5152 just in case I can make a backup of my own software and compare it with yours just in case, I'm not sure how to do a backup and restore with telnet.
backup I tried with the command
"cp -rf /net/mmx/mnt/app/eso/hmi/lsd/jars/lang_data.zip /fs/sda0"
but the command is not good,if you can write the correct telnet command
backup and restore.
thanks again

**Reply by filippz** [reply]
In my notes it seems that I was using something like this (please adjust it to your needs):

```
mount -uw /net/mmx/mnt/app
mount -uw /net/mmx/fs/sda0/

cp /net/mmx/mnt/app/eso/hmi/lsd/jars/lang_data.zip /net/mmx/fs/sda0/lang_data_MHI2_K3346.zip
cp /net/mmx/mnt/app/eso/hmi/lsd/jars/lang_data.zip /net/mmx/mnt/app/eso/hmi/lsd/jars/lang_data.zip.bak
cp /net/mmx/fs/sda0/lang_data_MHI2_K3346.zip /net/mmx/mnt/app/eso/hmi/lsd/jars/lang_data.zip

mount -ur /net/mmx/mnt/app
```

It would be probably easier to use SSH instead of SD card and transfer the files via SCP.

**Comment by DragosM90** [comment]
Hi @filippz, I translated the files from en_GB to ro_RO as you described to add Romanian to the MHI2 unit. I uploaded the new `lang_data.zip` to the unit via Wi-Fi using WinSCP, modified the `setlang` file to include ro_RO, and installed it on the MHI2 using the M.I.B. (More Incredible Bash) SD card; however, Romanian doesn't appear in the settings menu. The only way to get Romanian text on the MHI2 was by replacing the text within the en_US files with the Romanian translation. Consequently, the car's settings still show English (US), but when selected, the menu text appears in Romanian. This doesn't bother me too much, but I've run into a problem I can't seem to solve. I have a Virtual Cockpit, and because the car detects the MHI2 unit as set to en_US (even though the text is Romanian), the Virtual Cockpit doesn't display Romanian text—it stays in English. I used VCDS (module 17-Instruments) to set the language to Romanian; this changes the text to Romanian on the Virtual Cockpit as well, but after turning off the ignition and turning it back on, the Virtual Cockpit reverts to English. This makes me think there is a synchronization issue between the MHI2 and the Virtual Cockpit. Upon engine startup, the Virtual Cockpit checks the MHI2's language setting (which is en_US) and automatically switches to English. Can you help me get the Virtual Cockpit to display Romanian too? The Virtual Cockpit supports Romanian, but the setting doesn't stick...

Apologies for my English—it's not very good, and I'm using Google Translate; I hope you can understand what I mean.

Thanks!



<img width="1080" height="1920" alt="new-language-v0-ihzelri0eo7h1" src="https://github.com/user-attachments/assets/c741b685-86f4-4cb1-9a7c-56a3804a10d5" />

**Reply by filippz** [reply]
The "right" way would be to modify `setlang` script (https://github.com/Mr-MIBonk/M.I.B._More-Incredible-Bash/blob/main/apps/setlang#L35) to include `ro_RO` in the dataset for your particular firmware. The dataset format is described here: https://mibwiki.one/doc/language-dataset-7NoX2Q6jAc

**Reply by DragosM90** [reply]
I did everything exactly as you described above and the dataset as in the links above and the Romanian language still doesn't appear in the mhi2 menu, I don't understand what I'm doing wrong 😕

<img width="3213" height="5712" alt="IMG_8479" src="https://github.com/user-attachments/assets/04458a29-42cd-4980-a7d1-adc9252290ef" />
<img width="3213" height="5712" alt="IMG_8480" src="https://github.com/user-attachments/assets/6eb7c22f-ff9a-4367-a9c5-9cc53b8005ee" />
<img width="3213" height="5712" alt="IMG_8481" src="https://github.com/user-attachments/assets/2bd14994-1d53-46fd-aa3b-ed297394e30f" />
<img width="1080" height="1920" alt="att Dy4vmf4BWEe4a5KDp3tsFb5f1y9y3ZfDL1EYNo5IMw4" src="https://github.com/user-attachments/assets/477ec8c2-3706-4289-9b1e-bdd27f50f81f" />

**Reply by filippz** [reply]
I'll assume you're running MHI2_ER_AUG22? I have the same unit and I noticed that `setlang `in M.I.B. has different dataset DATASET_2D00 than my unit had originally. Please compare if `MU????-dataset.csv` from your backup has something different in `0xC0030006` than `setlang` uses. If so, I'd use that data as a base for `setlang` modification. Also, after adding `ro_RO` to the dataset you need to recalculate last two bytes (CRC-16/CCITT-False - you can use https://www.sunshine2k.de/coding/javascript/crc/crc_js.html - select CRC-16, CRC16_CCITT_FALSE, Input=Bytes - skip last two and replace them with the result)
I've newer actually added a language to the dataset so I might be wrong. In any case - please share your `setlang` to see if I would do something different based on mibwiki entry.

**Reply by DragosM90** [reply]
Yes, I'm using mhi2_er_aug22_k3344; where can I find the dataset (the setlang file) on my MHI2 unit—or is it on the M.I.B. card? That's where the dataset was located " DATASET_2D00=1364655F444503656E5F474203656E5F53410266725F46520169745F49540165735F45530170745F50540170745F4252016E6C5F4E4C0172755F525501706C5F504C0163735F435A0165735F4D5801656E5F55530173765F53450174725F54520268755F4855016E6F5F4E4F0161725F5341010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000003033AB9A " and that’s how I installed the new one. " 2064655F444503656E5F47420266725F46520169745F49540165735F4553016E6C5F4E4C0163735F435A0170745F50540173765F53450174725F54520272755F525501706C5F504C016E6F5F4E4F01656E5F55530165735F4D580170745F42520161725F53410262735F424100726F5F524F00736B5F534B00736C5F53490073725F52530064615F444B00656C5F47520068725F48520068755F48550066695F46490062675F4247006C765F4C560065745F4545006C745F4C5400756B5F554100656E5F474201656E5F474201656E5F474201656E5F474201656E5F474201656E5F474201656E5F474201656E5F474201656E5F474201656E5F474201656E5F474201656E5F474201656E5F474201656E5F474201656E5F474201656E5F474201656E5F474201656E5F474201656E5F47420163735F435A01656E5F474201656E5F474201656E5F474201656E5F474201656E5F474201656E5F474201656E5F474201656E5F474201656E5F474201656E5F474201656E5F47420172755F5255013233049D "


I have a backup of "appimg," but I can't find "setlang"...

**Reply by filippz** [reply]
M.I.B. should've made a backup folder on SD card and in a subfolder there should be `MU???-datasets.csv` file. Inside it there should be a original dataset for languages (line with 0xC0030006). Mine was/is:
`1364655f444503656e5f474203656e5f53410266725f46520169745f49540165735f45530170745f50540170745f4252016e6c5f4e4c0172755f525501706c5f504c0163735f435a0165735f4d5801656e5f55530173765f53450174725f54520268755f4855016e6f5f4e4f0161725f5341010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000003033ab9a`

I don't know why `ro_RO` is not being displayed by MMI, but I can try to just add `ro_RO` to dataset from my/your backup instead of using "MAX LANG" one from mibwiki and see what happens.

**Reply by filippz** [reply]
According to this https://github.com/NumberOneBot/mqb-mib-sound-datasets/blob/f23bfbfb6c6783bc413fd7f1f7565182ee3f9303/5F_MIB2-datasets.txt#L9 you should also add `ro_RO` to `/eso/hmi/lsd/lang_hmi_builtin.config`
This might brake something - but I guess you'll find out :)

**Reply by DragosM90** [reply]
I searched the entire card and can't find any datasets.csv. I'll try adding ro_RO to that config. I'll let you know if the unit still works after that 🤣

**Reply by DragosM90** [reply]
Honestly, I don't mind leaving the Romanian language instead of en_US, it bothers me that I can't find the command that the mib2 unit gives to the virtual cockpit, to be able to change that "path" from en_US to ro_RO. Do you have any idea where I can find the "path"?

Edit: I even replaced Portuguese with Romanian and when I select, mib2 is translated into Romanian but virtual cockpit changes it to Portuguese

**Reply by filippz** [reply]
> I searched the entire card and can't find any datasets.csv. I'll try adding ro_RO to that config. I'll let you know if the unit still works after that 🤣

Have you tried adding `ro_RO` to `/eso/hmi/lsd/lang_hmi_builtin.config`. If so, did it work?

**Reply by DragosM90** [reply]
I tried today, but nothing changed, I don't get an error on the screen, but nothing changes either 🙄

**Reply by filippz** [reply]
I've also tried today with no effect. I took a look at decompiled Java code from `lsd.jxe` and as far as I can tell the list of languages is constructed by intersecting of the languages defined in dataset, language list that Java loads from property called `languages.hmi.builtin` (both in `LanguageManager.java`) and `LANGUAGES` static list (defined in `I18NTarget.java` and contains `ro_RO`). The problem is that I'm not 100% certain that the dataset in `setlang` is the same dataset `LanguageManager.java` loads and I can't tell from where `languages.hmi.builtin` comes from. Logical assumption would be that we're talking about the same dataset and that `lang_hmi_builtin.config` somehow becomes `languages.hmi.builtin` Java property, but then it should be working and it's not so there is something more to figure out.
Onc thing that sticks out for me is that you have Romanian flag for English (I assume you're just changed translation of `en_GB`) and this confuses me even more - in a sense if the MMI is operating with `en_GB` setting how would it figure out that the flag to display is Romanian? I'm not seeing that flag in `/mnt/app/eso/hmi/lsd/images/` and I haven't figured out how filenames link to languages (for example we have `en_GB` and the GB flag is called `1100032.png`)

**Reply by DragosM90** [reply]
I'll try to explain each step I took:
1. I connected to the mhi2 drive with winscp
2. I extracted everything on the mhi2 drive (to have a complete back-up)
3. I copied lang_data.zip from the "back-up"
4. I unzipped lang_data.zip and deleted en_US from this archive from all folders (so I didn't touch en_GB, because I wanted to have English on the mhi2 drive and I thought it didn't make sense to have 2 English language options even if one is GB and one is US, so I gave up on US)
5. I put the Romanian language ro_RO from lang_data.zip (the one from mhi2q), but seeing that no matter what I did, the mhi2 drive wouldn't display the Romanian language, I thought of renaming ro_RO to en_US, so in lang_data.zip I have the Romanian language renamed to en_US
6. I I remade the .zip and injected it into the mhi2 unit with winscp
7. and surprise when I choose English (US, so not British) the text is displayed in Romanian
8. after a few days I thought about how to make the Romanian "flag" appear as well so I searched through the folders and found the folder with the "flags"
9. I opened the US flag (I don't know what numbers it had, but I know it's .png), then I edited it with photoshop, basically I drew the Romanian "flag" over the US "flag". (that's how the mhi2 unit ended up seeing the Romanian flag).

these were the main steps I took to have the Romanian language on the mhi2 unit, but I'm stuck on 2 things, next to the flag it says English, but I think I can solve that easily. The most annoying thing is that the unit knows that the language on the mhi2 unit is American English and sends this information to the Virtual Cockpit. I'm still trying to find this, where is the command that sends the mhi2 unit to the virtual cockpit that the language is American English. I want to find that command and change it to something like "my mhi2 has English, but I want you to display Romanian on the virtual cockpit). I specifically say that the Romanian text exists on the virtual cockpit because with both vcds and ODIS I can manually set the Romanian language from 17-instruments and it displays it, but this happens until I turn off the engine, turn on the ignition and when I turn on the ignition again the virtual cockpit returns to American English.

I hope you understand what I did and what I want to find (do)

Thank you for everything!

**Reply by DragosM90** [reply]
Look…
<img width="5712" height="3213" alt="IMG_8678" src="https://github.com/user-attachments/assets/6382e747-44aa-4fde-8177-5f10b340c45e" />

<img width="5712" height="3213" alt="IMG_8679" src="https://github.com/user-attachments/assets/50a8cb40-3be7-4391-b32d-704313ae147b" />

**Reply by DragosM90** [reply]
Watch until the end

https://youtu.be/ETSqzbE-gFk?is=AlgSR8LJmclYPQJu

**Reply by duxtodor-bit** [reply]
maybe the solution is simple, try simply installing mhi2q lang which has the Romanian language and if it works you have solved the problem

**Reply by DragosM90** [reply]
Don't you think I tried this the first time? If it were that simple, I wouldn't complicate things so much... did you even read what I wrote above? 🙄

**Reply by filippz** [reply]
After a lot of looking at the decompiled Java code I think that we have the following situation:

- languages are loaded from dataset (`setlang` can be used to write it - `0x2D00`)
- they are compared to `/eso/hmi/lsd/lang_hmi_builtin.config` (that somehow becomes `languages.hmi.builtin` Java property)
- MMI considers only languages in both lists
- if the selected language is not on the list the MMI reverts to "default" language (that is region dependent)
- at that point the current language is somehow pushed onto VC (my guess it's what is happening to you since your essentially using `en_US`)

I imagine that if you add `ro_RO` to dataset, `lang_hmi_builtin.config` and `lang_data.zip` you could use `setlang` to change `0x280` dataset to select `ro_RO` as current language. This should in theory allow for MMI to keep using `ro_RO` on startup and emit it as selected language and allow for VC to use `ro_RO` as the language. Also note that using `lang_data.zip` from any other firmware version will not works as different firmwares have different number of translation strings and they are _hardcoded_ in Java and must be kept in the same order for Java to load it properly.

The problem with selecting language from MMI is this:

- Java code has defined languages in `LANGUAGES` array in `de/audi/atip/i18n/I18NTarget.java` and `ro_RO` is in there
- languages for settings screen come from `de/audi/tghu/settings/hmi/evohigh/SettingsScreenBag1$1.java` in a sense that there are two arrays that use the same indexes:
  - one array containing flags (on my decompiled system there is an issue with endianess - so it's hard to figure out but essentially those numbers are filenames we're seeing in `app/eso/hmi/lsd/images/HMISettingsEvoHigh/`) - and there is no flag defined for `ro_RO` in that array so you can't see/select it
  - one array containing number representing indexes in `lang_data.zip` from `HMISettingsEvoHigh` folder which contain the text for that language (for example if you're using `en_GB` as MMI language the `hr_HR` language would be 'Croatian' while if you're using `de_DE` the same language would be 'Kroatisch')

In order to fix this we could:

- modify `de/audi/tghu/settings/hmi/evohigh/SettingsScreenBag1$1.java` to add `ro_RO` to the menu (there is `bs_BA` that needs to be added in between because we have to obey indexes defined in `LANGUAGES` array) so we add a number that will be it's flag
- then we need to add actual png to the `HMISettingsEvoHigh` folder
- we then must add a new line to every langauge in `lang_data.zip` `HMISettingsEvoHigh` with the actual name of the langauge in all those languages
- and add index position of that language to second array `de/audi/tghu/settings/hmi/evohigh/SettingsScreenBag1$1.java`

**Reply by DragosM90** [reply]
I'll try and let you know 😁

Do you have a full backup from an mhi2q drive?

thank you 🫡🙏

**Reply by filippz** [reply]
> Do you have a full backup from an mhi2q drive?

There are couple available at: https://mibsolution.one/#/1/14

**Reply by DragosM90** [reply]
I can't download from there, I don't know why 🤔

**Reply by filippz** [reply]
Well - this was actually quite complicated but I've made some progress:

- dataset from M.I.B. makes my MMI unable to use `en_GB` and `en_US` as TTS fails for those langauges (which also makes SDS unavailable) so I used the dataset from my backup and added `ro_RO` to it
- `/eso/hmi/lsd/lang_hmi_builtin.config` is not actually being used by `lsd.jxe`. `lsd.jxe` is technically ZIP that contains `resources/atip.properties` in which there is a definition of `languages.hmi.builtin` and could be updated
- in order not to mess with `lsd.jxe` I've patched/overriden `de.audi.tghu.lang.LangaugeManager` to add `ro_RO` after fetching the `languages.hmi.builtin`
- `de/audi/tghu/settings/hmi/evohigh/SettingsScreenBag1` and `de/audi/tghu/settings/hmi/evohigh/SettingsScreenBag2` are patched/overriden so they have a kind of map between flag numbers (`/mnt/app/eso/hmi/lsd/images/HMISettingsHigh`) and associated translation strings
- since `ro_RO` doesn't have associated TTS I've patched/overriden `de.audi.tghu.lang.LangaugeManager` to fall back to `en_GB` (otherwise it reverts to `de_DE`). In other words, navigation language is `en_GB`, and SDS is disabled when HMI langauge and TTS/navi have different values
- AudiMHI2Translator has been updated so it adds 3 new entries to every language in `lang_data.zip` when adding new language - one empty entry for flag number, and two string translations in `SettingsScreenBag1` and `SettingsScreenBag2`

Please note that `lang_data.zip` provided has already added `ro_RO` langauge but it's only a copy of `en_GB` and needs to be actually translated with `AudiMHI2Translator`. I needed to add that langauge in order to get/reserve number for png image file name and translations for `SettingsScreenBag1` and `SettingsScreenBag2`
Regarding the VC language, Inside `lsd.jxe` there is also a `de.audi.app.settings.timeunitslang.LanguageHandler` that translates indexes from `LANGUAGES` static list (defined in `de.audi.atip.i18n.I18NTarget`) to indexes that VC understands - `ro_RO` is mapped to 33, but my VC probably doesn't actually have support for `ro_RO` so it didn't work, it should work for your since your VC supports `ro_RO`.

**NEXT STEPS MIGHT BREAK YOUR UNIT - proceed at your own discretion**
Setup your unit by replacing/copying (do `mount -uw /mnt/app` before and `sync` after and keep backup of `lang_hmi_builtin.config` and `lang_data.zip`):
- `lang_hmi_builtin.config` to `/mnt/app/eso/hmi/lsd`
- `MU1438_AdditionalLanguages.jar` and `lang_data.zip` to `/mnt/app/eso/hmi/lsd/jars`
- png flag images to `/mnt/app/eso/hmi/lsd/images/HMISettingsHigh` 
- `setlang` to MIB SD card (app folder)

Execute `setlang` from `MIB` and keep your fingers crossed.

Final result:
<img width="1024" height="768" alt="image" src="https://github.com/user-attachments/assets/a796c55f-130e-4668-b63d-9718aabaad01" />
[MHI2_ER_AUG22_K3346_rev1.zip](https://github.com/user-attachments/files/32149044/MHI2_ER_AUG22_K3346_rev1.zip)

**Reply by DragosM90** [reply]
Thank you very much!

Everything is working now, both on the main screen and on the virtual cockpit 🙏🙏🙏

<img width="5712" height="3213" alt="IMG_8967" src="https://github.com/user-attachments/assets/df057df2-5958-4dd1-9fe7-62035463988f" />

**Reply by filippz** [reply]
Cool!

Please share your `lang_data.zip` when/if you're done with the translation to Romanian.

**Reply by DragosM90** [reply]
Sure! On the virtual cockpit, to have the Romanian language, you only need a firmware update, there I didn't need to translate anything, after the firmware update the Romanian language is available as OEM

**Reply by DragosM90** [reply]
sorry for the delay, I made a few more adjustments to the translation and I still have a few lines to translate, I'll add the rest of the changes as time allows

lang_data - language romanian
https://mega.nz/file/wyBmhIhY#unYW9hLfgDnFjZQ6n4pipno4wxSo2R_xx0UujCWVgZw

**Reply by filippz** [reply]
You seem to be using old `lang_data.zip` for translation - the one I've sent in `MHI2_ER_AUG22_K3346_rev1.zip` has some more translation strings so additional language names are visible in the settings menu. My guess is that you just have Romanian flag and not string "ro_RO" next to it in the settings?

Anyway - I'll add those additional strings once you're done translating or I can add them to the `lang_data.zip` you've already sent and you can continue translating after I send you the fixed `lang_data.zip`?

<img width="732" height="1125" alt="image" src="https://github.com/user-attachments/assets/a135ea8f-3944-4a2c-9f0a-4d56f1e5f4c4" />

**Reply by DragosM90** [reply]
Yes, I used the old file, but I also have the Romanian flag and the text “Romana” next to the flag. What you gave me appeared next to the flag “ro_RO”

**Reply by filippz** [reply]
I don't know from where could "Romana" text come from - I don't see it in your file and I also can't find it in Java code. I was lazy and "my" `data_lang.zip` just added entries 1494 and 1495 as "ro_RO" - they both should be "Romana".
In any case, when you're done I'll copy your translations to "my" `data_lang.zip` and extend it with entries 1490-1516 for other langauges and you'll retest with that file to se if everything is still working as expected. If someone else decides to add another language we'll add it to that file - otherwise we could end up with many different versions of `data_lang.zip`. We should try to keep one `data_lang.zip` for one firmware.

**Reply by DragosM90** [reply]
reads it from here:

<img width="969" height="597" alt="image" src="https://github.com/user-attachments/assets/8bf380c8-139a-4063-bc06-0e154aa03f1c" />

<img width="695" height="847" alt="image" src="https://github.com/user-attachments/assets/6136981f-7d17-4045-bb92-07cded1ec680" />

In my old file for "Romana" it was listed as "Suomi"—that’s what appeared next to the Romanian flag; I searched for "Suomi," changed it to "Romana" and—lo and behold—"Romana" showed up in the MMI settings next to the little flag.

<img width="639" height="819" alt="image" src="https://github.com/user-attachments/assets/ab3696d2-1989-45df-a7de-bb7489f08c62" />

When I get to the car, I'll post the photo to show you

**Reply by filippz** [reply]
We are not using same `data_lang.zip` as a base.
To check I've extracted `data_lang.zip` from K3346 (K3344 uses the exact same file) and P3241 and in `HMISettingsEvoHigh` part they both end on index 1489, while older K2161 ends on index 1390.
Your `data_lang.zip` ends on index 1512 so my guess is you have somehow started to use lang_data.zip from Quallcom MHI2Q based units because I'm seeing exactly the same ending index of 1512 in those.
**Please switch to my `data_lang.zip`** because index positions are how Java code fetches appropriate string/translations and you get messed up text if you use the wrong `data_lang.zip`. That's why you saw "_Suomi_" on one of settings screen - on the other you would get "_Slovenščina_" as I'm using indexes 1494 and 1495 for "_main_" settings screen and "_language_" settings screen.

**Reply by DragosM90** [reply]
Ok, I'll do as you say, but is there no problem that I have k3344 and you have k3346?

**Reply by filippz** [reply]
Both K3344 and K3346 have exactly the same `data_lang.zip` (exactly the same `app.img` - which means the same Java code, translations,...) so you can continue using K3344.
They've just done something to the radio tuner parts as far as I can tell:
<img width="1435" height="1251" alt="image" src="https://github.com/user-attachments/assets/e8f509f9-a4a7-445e-949f-618bf8f1f895" />

**Reply by DragosM90** [reply]
I have a patch installed that displays waze and google maps on the virtual cockpit and if I update I feel like I might "lose" it 😅

**Reply by filippz** [reply]
You might try to start the update - I think update will just select components that need updating (marked with Y) and leave everything else as is. On the other hand nothing really changed so maybe it's best not to risk breaking your additions.

**Comment by dux72dc** [comment]
hi filippz,
This is my original version of MHI2Q_5152; could you check if it is the same as MHI2Q_10? It seems to me there are some differences in certain lines, or perhaps you have a tool—like a translation tool—that would allow me to compare the same language (e.g., British English) across two different language files?

[lang_data.zip](https://github.com/user-attachments/files/32644776/lang_data.zip)

**Reply by filippz** [reply]
Your file is identical to the one I have in my development folder named `lang_data_MHI2Q_10.zip`. I'm not sure where I actually got it but I had an idea to use `si_SI` langauage from it to do `hr_HR`.
Anyway - app allows for an reference mode as described here: [#discussioncomment-18076442](#discussioncomment-18076442)
Since different firmwares have different strings it's expected that not all strings have matches across different firmwares. Nevertheless, a lot of strings do match and it could speed up translation - you cane either accept offered translation (via "Use translation" checkbox) or do a manual translation of every string.
